File: core/queue_manager.py
"""
QueueManager — Async queue wrapper for music tracks.
"""

from __future__ import annotations

import asyncio
from typing import Optional
from core.ytdl_source import Track


class QueueManager:
    """Thread-safe async queue for managing music tracks."""

    def __init__(self):
        self._queue: list[Track] = []
        self._lock = asyncio.Lock()
        self._counter = 0

    async def add(self, track: Track) -> int:
        """Add a track to the queue. Returns queue position."""
        async with self._lock:
            self._counter += 1
            track.insert_id = self._counter
            self._queue.append(track)
            return len(self._queue)

    async def get_next(self) -> Optional[Track]:
        """Get and remove the next track from the queue."""
        async with self._lock:
            if self._queue:
                return self._queue.pop(0)
            return None

    async def clear(self):
        """Clear all tracks from the queue."""
        async with self._lock:
            self._queue.clear()
            self._counter = 0

    async def remove(self, index: int) -> Optional[Track]:
        """Remove a track at a specific index (0-based). Returns removed track."""
        async with self._lock:
            if 0 <= index < len(self._queue):
                return self._queue.pop(index)
            return None

    async def shuffle(self, mode: int):
        """
        Shuffle the queue based on mode:
        0: Off (Restore original order)
        1: Standard (Random)
        2: Riffle (Interleave)
        """
        import random
        async with self._lock:
            if mode == 0:
                # Restore original order based on insert_id
                self._queue.sort(key=lambda t: t.insert_id)
            elif mode == 1:
                # Standard random shuffle
                random.shuffle(self._queue)
            elif mode == 2:
                # Riffle shuffle (Simulate card shuffling)
                # Split into two halves and interleave
                if len(self._queue) < 2:
                    return

                # Perform a few riffs for better effect
                for _ in range(3):
                    mid = len(self._queue) // 2
                    left = self._queue[:mid]
                    right = self._queue[mid:]
                    self._queue = []
                    while left or right:
                        if left: self._queue.append(left.pop(0))
                        if right: self._queue.append(right.pop(0))

    @property
    def is_empty(self) -> bool:
        """Check if queue is empty."""
        return len(self._queue) == 0

    @property
    def size(self) -> int:
        """Get number of tracks in queue."""
        return len(self._queue)

    def as_list(self, limit: int = 10) -> list[Track]:
        """Get a copy of the queue as a list, limited to `limit` items."""
        return self._queue[:limit]

    async def put_front(self, track: Track):
        """Put a track at the front of the queue (for loop single)."""
        async with self._lock:
            self._queue.insert(0, track)

    async def put_back(self, track: Track):
        """Put a track at the back of the queue (for loop queue)."""
        async with self._lock:
            self._queue.append(track)

File: src/player/queueManager.js
export class QueueManager {
  constructor() {
    this.queue = [];
    this.counter = 0;
  }

  add(track) {
    this.counter += 1;
    track.insertId = this.counter;
    this.queue.push(track);
    return this.queue.length;
  }

  getNext() {
    return this.queue.shift() ?? null;
  }

  peekNext() {
    return this.queue[0] ?? null;
  }

  clear() {
    this.queue = [];
    this.counter = 0;
  }

  remove(index) {
    if (index < 0 || index >= this.queue.length) {
      return null;
    }
    return this.queue.splice(index, 1)[0] ?? null;
  }

  move(sourceIndex, targetIndex) {
    if (sourceIndex < 0 || sourceIndex >= this.queue.length) {
      return null;
    }
    const clampedTarget = Math.max(0, Math.min(targetIndex, this.queue.length - 1));
    const [track] = this.queue.splice(sourceIndex, 1);
    this.queue.splice(clampedTarget, 0, track);
    return track;
  }

  shuffle(mode) {
    if (mode === 0) {
      this.queue.sort((a, b) => a.insertId - b.insertId);
      return;
    }
    if (mode === 1) {
      for (let i = this.queue.length - 1; i > 0; i -= 1) {
        const j = Math.floor(Math.random() * (i + 1));
        [this.queue[i], this.queue[j]] = [this.queue[j], this.queue[i]];
      }
      return;
    }
    if (mode === 2 && this.queue.length > 1) {
      for (let round = 0; round < 3; round += 1) {
        const middle = Math.floor(this.queue.length / 2);
        const left = this.queue.slice(0, middle);
        const right = this.queue.slice(middle);
        const mixed = [];
        while (left.length || right.length) {
          if (left.length) mixed.push(left.shift());
          if (right.length) mixed.push(right.shift());
        }
        this.queue = mixed;
      }
    }
  }

  putFront(track) {
    this.queue.unshift(track);
  }

  putBack(track) {
    this.queue.push(track);
  }

  prune(predicate) {
    const kept = [];
    const removed = [];
    for (const track of this.queue) {
      try {
        if (predicate(track)) {
          kept.push(track);
        } else {
          removed.push(track);
        }
      } catch {
        removed.push(track);
      }
    }
    this.queue = kept;
    return removed;
  }

  asList(limit = 20) {
    return this.queue.slice(0, limit);
  }

  get size() {
    return this.queue.length;
  }

  get isEmpty() {
    return this.queue.length === 0;
  }
}
